// @eth-optimism/contracts-bedrock/dist/types/contracts/deployment/MigrationSystemDictator.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  Overrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type {
  FunctionFragment,
  Result,
  EventFragment,
} from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
  PromiseOrValue,
} from "../../common";

export declare namespace BaseSystemDictator {
  export type GlobalConfigStruct = {
    addressManager: PromiseOrValue<string>;
    proxyAdmin: PromiseOrValue<string>;
    controller: PromiseOrValue<string>;
    finalOwner: PromiseOrValue<string>;
  };

  export type GlobalConfigStructOutput = [string, string, string, string] & {
    addressManager: string;
    proxyAdmin: string;
    controller: string;
    finalOwner: string;
  };

  export type ProxyAddressConfigStruct = {
    l2OutputOracleProxy: PromiseOrValue<string>;
    optimismPortalProxy: PromiseOrValue<string>;
    l1CrossDomainMessengerProxy: PromiseOrValue<string>;
    l1StandardBridgeProxy: PromiseOrValue<string>;
    optimismMintableERC20FactoryProxy: PromiseOrValue<string>;
    l1ERC721BridgeProxy: PromiseOrValue<string>;
  };

  export type ProxyAddressConfigStructOutput = [
    string,
    string,
    string,
    string,
    string,
    string
  ] & {
    l2OutputOracleProxy: string;
    optimismPortalProxy: string;
    l1CrossDomainMessengerProxy: string;
    l1StandardBridgeProxy: string;
    optimismMintableERC20FactoryProxy: string;
    l1ERC721BridgeProxy: string;
  };

  export type ImplementationAddressConfigStruct = {
    l2OutputOracleImpl: PromiseOrValue<string>;
    optimismPortalImpl: PromiseOrValue<string>;
    l1CrossDomainMessengerImpl: PromiseOrValue<string>;
    l1StandardBridgeImpl: PromiseOrValue<string>;
    optimismMintableERC20FactoryImpl: PromiseOrValue<string>;
    l1ERC721BridgeImpl: PromiseOrValue<string>;
    portalSenderImpl: PromiseOrValue<string>;
  };

  export type ImplementationAddressConfigStructOutput = [
    string,
    string,
    string,
    string,
    string,
    string,
    string
  ] & {
    l2OutputOracleImpl: string;
    optimismPortalImpl: string;
    l1CrossDomainMessengerImpl: string;
    l1StandardBridgeImpl: string;
    optimismMintableERC20FactoryImpl: string;
    l1ERC721BridgeImpl: string;
    portalSenderImpl: string;
  };

  export type L2OutputOracleConfigStruct = {
    l2OutputOracleGenesisL2Output: PromiseOrValue<BytesLike>;
    l2OutputOracleProposer: PromiseOrValue<string>;
    l2OutputOracleOwner: PromiseOrValue<string>;
  };

  export type L2OutputOracleConfigStructOutput = [string, string, string] & {
    l2OutputOracleGenesisL2Output: string;
    l2OutputOracleProposer: string;
    l2OutputOracleOwner: string;
  };

  export type SystemConfigStruct = {
    globalConfig: BaseSystemDictator.GlobalConfigStruct;
    proxyAddressConfig: BaseSystemDictator.ProxyAddressConfigStruct;
    implementationAddressConfig: BaseSystemDictator.ImplementationAddressConfigStruct;
    l2OutputOracleConfig: BaseSystemDictator.L2OutputOracleConfigStruct;
  };

  export type SystemConfigStructOutput = [
    BaseSystemDictator.GlobalConfigStructOutput,
    BaseSystemDictator.ProxyAddressConfigStructOutput,
    BaseSystemDictator.ImplementationAddressConfigStructOutput,
    BaseSystemDictator.L2OutputOracleConfigStructOutput
  ] & {
    globalConfig: BaseSystemDictator.GlobalConfigStructOutput;
    proxyAddressConfig: BaseSystemDictator.ProxyAddressConfigStructOutput;
    implementationAddressConfig: BaseSystemDictator.ImplementationAddressConfigStructOutput;
    l2OutputOracleConfig: BaseSystemDictator.L2OutputOracleConfigStructOutput;
  };
}

export interface MigrationSystemDictatorInterface extends utils.Interface {
  functions: {
    "config()": FunctionFragment;
    "currentStep()": FunctionFragment;
    "owner()": FunctionFragment;
    "renounceOwnership()": FunctionFragment;
    "step1()": FunctionFragment;
    "step2()": FunctionFragment;
    "step3()": FunctionFragment;
    "step4()": FunctionFragment;
    "step5()": FunctionFragment;
    "step6()": FunctionFragment;
    "transferOwnership(address)": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic:
      | "config"
      | "currentStep"
      | "owner"
      | "renounceOwnership"
      | "step1"
      | "step2"
      | "step3"
      | "step4"
      | "step5"
      | "step6"
      | "transferOwnership"
  ): FunctionFragment;

  encodeFunctionData(functionFragment: "config", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "currentStep",
    values?: undefined
  ): string;
  encodeFunctionData(functionFragment: "owner", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "renounceOwnership",
    values?: undefined
  ): string;
  encodeFunctionData(functionFragment: "step1", values?: undefined): string;
  encodeFunctionData(functionFragment: "step2", values?: undefined): string;
  encodeFunctionData(functionFragment: "step3", values?: undefined): string;
  encodeFunctionData(functionFragment: "step4", values?: undefined): string;
  encodeFunctionData(functionFragment: "step5", values?: undefined): string;
  encodeFunctionData(functionFragment: "step6", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "transferOwnership",
    values: [PromiseOrValue<string>]
  ): string;

  decodeFunctionResult(functionFragment: "config", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "currentStep",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "owner", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "renounceOwnership",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "step1", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "step2", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "step3", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "step4", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "step5", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "step6", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "transferOwnership",
    data: BytesLike
  ): Result;

  events: {
    "OwnershipTransferred(address,address)": EventFragment;
  };

  getEvent(nameOrSignatureOrTopic: "OwnershipTransferred"): EventFragment;
}

export interface OwnershipTransferredEventObject {
  previousOwner: string;
  newOwner: string;
}
export type OwnershipTransferredEvent = TypedEvent<
  [string, string],
  OwnershipTransferredEventObject
>;

export type OwnershipTransferredEventFilter =
  TypedEventFilter<OwnershipTransferredEvent>;

export interface MigrationSystemDictator extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: MigrationSystemDictatorInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    config(
      overrides?: CallOverrides
    ): Promise<
      [
        BaseSystemDictator.GlobalConfigStructOutput,
        BaseSystemDictator.ProxyAddressConfigStructOutput,
        BaseSystemDictator.ImplementationAddressConfigStructOutput,
        BaseSystemDictator.L2OutputOracleConfigStructOutput
      ] & {
        globalConfig: BaseSystemDictator.GlobalConfigStructOutput;
        proxyAddressConfig: BaseSystemDictator.ProxyAddressConfigStructOutput;
        implementationAddressConfig: BaseSystemDictator.ImplementationAddressConfigStructOutput;
        l2OutputOracleConfig: BaseSystemDictator.L2OutputOracleConfigStructOutput;
      }
    >;

    currentStep(overrides?: CallOverrides): Promise<[number]>;

    owner(overrides?: CallOverrides): Promise<[string]>;

    renounceOwnership(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    step1(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    step2(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    step3(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    step4(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    step5(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    step6(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;

    transferOwnership(
      newOwner: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;
  };

  config(
    overrides?: CallOverrides
  ): Promise<
    [
      BaseSystemDictator.GlobalConfigStructOutput,
      BaseSystemDictator.ProxyAddressConfigStructOutput,
      BaseSystemDictator.ImplementationAddressConfigStructOutput,
      BaseSystemDictator.L2OutputOracleConfigStructOutput
    ] & {
      globalConfig: BaseSystemDictator.GlobalConfigStructOutput;
      proxyAddressConfig: BaseSystemDictator.ProxyAddressConfigStructOutput;
      implementationAddressConfig: BaseSystemDictator.ImplementationAddressConfigStructOutput;
      l2OutputOracleConfig: BaseSystemDictator.L2OutputOracleConfigStructOutput;
    }
  >;

  currentStep(overrides?: CallOverrides): Promise<number>;

  owner(overrides?: CallOverrides): Promise<string>;

  renounceOwnership(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  step1(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  step2(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  step3(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  step4(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  step5(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  step6(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  transferOwnership(
    newOwner: PromiseOrValue<string>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  callStatic: {
    config(
      overrides?: CallOverrides
    ): Promise<
      [
        BaseSystemDictator.GlobalConfigStructOutput,
        BaseSystemDictator.ProxyAddressConfigStructOutput,
        BaseSystemDictator.ImplementationAddressConfigStructOutput,
        BaseSystemDictator.L2OutputOracleConfigStructOutput
      ] & {
        globalConfig: BaseSystemDictator.GlobalConfigStructOutput;
        proxyAddressConfig: BaseSystemDictator.ProxyAddressConfigStructOutput;
        implementationAddressConfig: BaseSystemDictator.ImplementationAddressConfigStructOutput;
        l2OutputOracleConfig: BaseSystemDictator.L2OutputOracleConfigStructOutput;
      }
    >;

    currentStep(overrides?: CallOverrides): Promise<number>;

    owner(overrides?: CallOverrides): Promise<string>;

    renounceOwnership(overrides?: CallOverrides): Promise<void>;

    step1(overrides?: CallOverrides): Promise<void>;

    step2(overrides?: CallOverrides): Promise<void>;

    step3(overrides?: CallOverrides): Promise<void>;

    step4(overrides?: CallOverrides): Promise<void>;

    step5(overrides?: CallOverrides): Promise<void>;

    step6(overrides?: CallOverrides): Promise<void>;

    transferOwnership(
      newOwner: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<void>;
  };

  filters: {
    "OwnershipTransferred(address,address)"(
      previousOwner?: PromiseOrValue<string> | null,
      newOwner?: PromiseOrValue<string> | null
    ): OwnershipTransferredEventFilter;
    OwnershipTransferred(
      previousOwner?: PromiseOrValue<string> | null,
      newOwner?: PromiseOrValue<string> | null
    ): OwnershipTransferredEventFilter;
  };

  estimateGas: {
    config(overrides?: CallOverrides): Promise<BigNumber>;

    currentStep(overrides?: CallOverrides): Promise<BigNumber>;

    owner(overrides?: CallOverrides): Promise<BigNumber>;

    renounceOwnership(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    step1(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    step2(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    step3(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    step4(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    step5(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    step6(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;

    transferOwnership(
      newOwner: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    config(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    currentStep(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    owner(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    renounceOwnership(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    step1(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    step2(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    step3(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    step4(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    step5(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    step6(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;

    transferOwnership(
      newOwner: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;
  };
}
